File: scripts/crawl-shuttle-bus.ts
/**
 * 스쿨버스 시간표 크롤링 스크립트
 * https://www.syu.ac.kr/school-life/school-bus/
 */

import axios from "axios";
import * as cheerio from "cheerio";
import * as fs from "fs";
import * as path from "path";

interface ShuttleBusSchedule {
  id: string;
  routeName: string;
  startLocation: string;
  endLocation: string;
  schedules: {
    weekday: string[];
    weekend: string[];
  };
  lastUpdated: string;
}

async function crawlShuttleBusSchedule() {
  const schedules: ShuttleBusSchedule[] = [];
  const baseUrl = "https://www.syu.ac.kr/school-life/school-bus/";

  console.log("🚌 스쿨버스 시간표 크롤링 시작...");

  try {
    const response = await axios.get(baseUrl, {
      headers: {
        "User-Agent":
          "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
      },
    });

    const $ = cheerio.load(response.data);

    // 테이블에서 시간표 추출
    // 구조: 테이블의 각 행이 하나의 노선
    let routeIndex = 0;

    // 모든 테이블 찾기
    $("table").each((_tableIdx, tableElement) => {
      const $table = $(tableElement);

      // 테이블 제목 찾기 (평일/주말)
      const tableTitle = $table.prev("h3, h4, strong").text().trim();
      const isWeekday = tableTitle.includes("평일");
      const isWeekend = tableTitle.includes("주말");

      // 테이블 행 반복
      $table.find("tbody tr").each((_rowIdx, trElement) => {
        const $row = $(trElement);
        const cells = $row.find("td");

        if (cells.length >= 2) {
          // 첫 번째 셀: 노선명
          const routeName = $(cells[0]).text().trim();
          // 두 번째 셀: 시간표 (시간들이 쉼표로 구분되거나 개별 셀)
          const timeText = $(cells[1]).text().trim();

          if (routeName && timeText) {
            // 기존 노선이 있는지 확인
            let existingRoute = schedules.find((s) =>
              s.routeName.includes(routeName.split("-")[0].trim()),
            );

            if (!existingRoute) {
              existingRoute = {
                id: `shuttle-${++routeIndex}`,
                routeName: routeName,
                startLocation: "캠퍼스",
                endLocation: routeName.split("-")[1]?.trim() || "목적지",
                schedules: {
                  weekday: [],
                  weekend: [],
                },
                lastUpdated: new Date().toISOString().split("T")[0],
              };
              schedules.push(existingRoute);
            }

            // 시간 파싱
            const times = timeText
              .split(/[,\s]+/)
              .filter((t) => /^\d{1,2}:\d{2}/.test(t))
              .map((t) => t.substring(0, 5)); // HH:MM 형식

            if (isWeekday) {
              existingRoute.schedules.weekday = [
                ...new Set([...existingRoute.schedules.weekday, ...times]),
              ]
                .sort()
                .slice(0, 7); // 평일 최대 7개 시간
            }

            if (isWeekend) {
              existingRoute.schedules.weekend = [
                ...new Set([...existingRoute.schedules.weekend, ...times]),
              ]
                .sort()
                .slice(0, 5); // 주말 최대 5개 시간
            }
          }
        }
      });
    });

    // 모든 tr 추출 및 시간표 처리
    $("table tbody tr").each((_index, element) => {
      const $row = $(element);
      const cells = $row.find("td");

      if (cells.length >= 3) {
        const routeName = $(cells[0]).text().trim();
        const destination = $(cells[1]).text().trim();

        // 평일/주말 시간표를 각 열에서 추출
        const getTimeSlots = (startIndex: number): string[] => {
          const slots: string[] = [];
          for (let i = startIndex; i < cells.length; i++) {
            const time = $(cells[i]).text().trim();
            if (/^\d{1,2}:\d{2}/.test(time)) {
              slots.push(time.substring(0, 5));
            }
          }
          return slots;
        };

        // 새로운 노선 추가
        if (routeName && !schedules.some((s) => s.routeName === routeName)) {
          const weekdayTimes = getTimeSlots(2);
          const weekendTimes = getTimeSlots(Math.ceil(cells.length / 2) + 2);

          schedules.push({
            id: `shuttle-${schedules.length + 1}`,
            routeName: routeName,
            startLocation: "캠퍼스",
            endLocation: destination,
            schedules: {
              weekday: weekdayTimes.slice(0, 7), // 최대 7개
              weekend: weekendTimes.slice(0, 5), // 최대 5개
            },
            lastUpdated: new Date().toISOString().split("T")[0],
          });
        }
      }
    });

    // 데이터가 너무 적으면 기본값으로 채우기
    if (schedules.length < 2) {
      console.log("📌 공개 페이지에서 충분한 데이터를 찾지 못했습니다.");
      console.log("   기본 예시 데이터로 저장합니다.");

      schedules.push(
        {
          id: "shuttle-001",
          routeName: "신복로 - 종로",
          startLocation: "캠퍼스 정주교",
          endLocation: "롯데마트 종로점",
          schedules: {
            weekday: [
              "06:30",
              "07:00",
              "07:30",
              "08:00",
              "12:00",
              "12:30",
              "13:00",
              "17:30",
              "18:00",
              "18:30",
            ],
            weekend: ["08:00", "09:00", "10:00", "12:00", "14:00"],
          },
          lastUpdated: new Date().toISOString().split("T")[0],
        },
        {
          id: "shuttle-002",
          routeName: "신복로 - 강남역",
          startLocation: "캠퍼스 정주교",
          endLocation: "강남역 11번 출구",
          schedules: {
            weekday: [
              "06:45",
              "07:15",
              "07:45",
              "08:15",
              "12:15",
              "13:15",
              "17:45",
              "18:15",
              "18:45",
              "19:15",
            ],
            weekend: ["08:30", "09:30", "10:30", "12:30", "14:30"],
          },
          lastUpdated: new Date().toISOString().split("T")[0],
        },
        {
          id: "shuttle-003",
          routeName: "신복로 - 동대문",
          startLocation: "캠퍼스 정주교",
          endLocation: "동대문역 13번 출구",
          schedules: {
            weekday: [
              "07:00",
              "07:30",
              "08:00",
              "08:30",
              "12:30",
              "13:30",
              "18:00",
              "18:30",
              "19:00",
              "19:30",
            ],
            weekend: ["09:00", "10:00", "11:00", "13:00", "15:00"],
          },
          lastUpdated: new Date().toISOString().split("T")[0],
        },
      );
    }

    // JSON 파일로 저장
    const dataPath = path.join(
      process.cwd(),
      "public",
      "data",
      "shuttle-bus-schedule.json",
    );

    fs.mkdirSync(path.dirname(dataPath), { recursive: true });
    fs.writeFileSync(dataPath, JSON.stringify(schedules, null, 2), "utf-8");

    console.log(
      `✅ 스쿨버스 시간표 ${schedules.length}개 노선 저장 완료: ${dataPath}`,
    );
  } catch (error) {
    console.error("❌ 스쿨버스 시간표 크롤링 실패:", error);
    throw error;
  }
}

// 실행
crawlShuttleBusSchedule();

File: scripts/crawl_shuttle_bus.py
# -*- coding: utf-8 -*-
import sys
import io
sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8')
"""
셔틀버스 시간표 크롤링 스크립트 (변경사항만 반영)
월 1회 실행됨 - 기존 데이터와 비교 후 변경된 노선만 업데이트
https://www.syu.ac.kr/school-life/school-bus/
"""

import requests
from bs4 import BeautifulSoup
import json
import os
import re
from datetime import datetime

def crawl_shuttle_bus():
    """셔틀버스 시간표 크롤링"""
    data_path = "public/data/shuttle-bus-schedule.json"
    
    # 기존 데이터 로드
    existing_routes = {}
    if os.path.exists(data_path):
        try:
            with open(data_path, 'r', encoding='utf-8') as f:
                existing_data = json.load(f)
                for route in existing_data:
                    existing_routes[route['routeName']] = route
                print(f"📌 기존 노선 {len(existing_routes)}개 로드됨")
        except Exception as e:
            print(f"⚠️  기존 데이터 로드 실패: {e}")
    
    url = "https://www.syu.ac.kr/school-life/school-bus/"
    headers = {
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36"
    }
    
    print("🚌 셔틀버스 시간표 크롤링 시작...")
    
    try:
        response = requests.get(url, headers=headers, timeout=10)
        response.encoding = 'utf-8'
        
        if response.status_code != 200:
            print(f"❌ 요청 실패: {response.status_code}")
            return
        
        soup = BeautifulSoup(response.text, 'html.parser')
        
        schedules = []
        tables = soup.find_all("table")
        print(f"📊 총 {len(tables)}개의 테이블 발견")
        
        for table in tables:
            try:
                # 제목 찾기
                header_title = ""
                prev_elem = table.find_previous(["h3", "h4", "strong", "b", "p"])
                if prev_elem:
                    header_title = prev_elem.get_text(strip=True)
                
                # 목적지 찾기
                route_title = ""
                header_row = table.select_one("thead tr")
                if header_row:
                    route_title = header_row.get_text(strip=True)
                
                # 첫 행에서 목적지 추출
                first_row = table.select_one("tbody tr")
                if not first_row:
                    continue
                
                cells = first_row.select("td")
                if not cells:
                    continue
                
                destination = cells[0].get_text(strip=True) or route_title or header_title
                
                # 시간 추출
                times = []
                for tr in table.select("tbody tr"):
                    for td in tr.select("td"):
                        time_text = td.get_text(strip=True)
                        if re.match(r'^\d{1,2}:\d{2}', time_text):
                            time = time_text[:5]
                            if time not in times:
                                times.append(time)
                
                if times:
                    times.sort()
                    
                    # 금요일 여부 확인
                    is_alt_day = "금" in (header_title + route_title)
                    
                    schedule = {
                        "id": f"shuttle-{len(schedules)}",
                        "routeName": destination,
                        "startLocation": "삼육대학교",
                        "endLocation": destination,
                        "schedules": {
                            "weekday": times if not is_alt_day else [],
                            "weekdayAlt": times if is_alt_day else [],
                            "weekend": []
                        },
                        "lastUpdated": datetime.now().isoformat()
                    }
                    
                    # 기존 데이터와 비교
                    if destination in existing_routes:
                        old_times = existing_routes[destination]['schedules'].get('weekday', [])
                        if old_times != times:
                            print(f"  🔄 노선 업데이트: {destination}")
                        else:
                            print(f"  ✓ 노선 유지: {destination}")
                        # 기존 데이터의 다른 필드 유지
                        schedule = {**existing_routes[destination], **schedule}
                    else:
                        print(f"  ✨ 신규 노선: {destination}")
                    
                    schedules.append(schedule)
            
            except Exception as e:
                print(f"  ⚠️  테이블 파싱 오류: {e}")
                continue
        
        # 디렉토리 생성 및 JSON 파일로 저장
        os.makedirs(os.path.dirname(data_path), exist_ok=True)
        with open(data_path, 'w', encoding='utf-8') as f:
            json.dump(schedules, f, ensure_ascii=False, indent=2)
        
        print(f"✅ 셔틀버스 {len(schedules)}개 노선 저장 완료")
    
    except Exception as e:
        print(f"❌ 셔틀버스 크롤링 실패: {e}")
        import traceback
        traceback.print_exc()
        raise

if __name__ == "__main__":
    crawl_shuttle_bus()
